input/drivers_hid/iohidmanager_hid.c

```diff
@@ -476,14 +474,43 @@ static void iohidmanager_hid_device_input_callback(void *data, IOReturn result,
       }
    }
 }
-
-static void iohidmanager_hid_device_remove(void *data,
-      IOReturn result, void* sender)
+static void list_controllers()
 {
-   struct iohidmanager_hid_adapter *adapter =
-      (struct iohidmanager_hid_adapter*)data;
-   iohidmanager_hid_t *hid = (iohidmanager_hid_t*)
-      hid_driver_get_data();
+   iohidmanager_hid_t *hid = (iohidmanager_hid_t*)   hid_driver_get_data();
+   if (!hid)
+      return;
+   for (int i=0; i<4; i++)
+   {
+      struct iohidmanager_hid_adapter *a = (struct iohidmanager_hid_adapter*)hid->slots[i].data;
+      if (!a)
+      {
+         printf("Port %d is not assigned to a controller on HID %p \n", i, hid);
+         continue;
+      }
+      else
+         printf("Port %d is assigned to device %p on HID %p \n", i, a->handle, hid);
+   }
+   printf("\n");
+}
+static void iohidmanager_hid_device_remove(IOHIDDeviceRef device, iohidmanager_hid_t* hid)
+{
+   struct iohidmanager_hid_adapter *adapter = NULL;
+   //loop though the controller ports and find the device with a matching IOHINDeviceRef
+   for (int i=0; i<MAX_USERS; i++)
+   {
+      struct iohidmanager_hid_adapter *a = (struct iohidmanager_hid_adapter*)hid->slots[i].data;
+      if (!a)
+         continue;
+      if (a->handle == device)
+      {
+         adapter = a;
+         break;
+      }
+   }
+   if (!adapter)
+   {
+      printf("Error removing device %p from HID %p\n",device, hid);
```

Comment:
Are all these printfs necessary? We could replace them with RARCH_LOG/RARCH_ERR/RARCH_WARN if they are truly essential. Do note that the logging code is not thread-safe so if this function happens on a thread, perhaps not logging at all would be the safest choice. But I'll leave this determination up to you to decide.

Reply:
The printfs aren't necessary, but I was reluctant to remove them yet as they may still be useful as things get tested.

However, I think *some* additional logs would be nice. As of right now the *only* log that's output is when a controller is connected. I think we at least have logs when a controller is disconnected as well as a log detailing whether the device was matched by the autoconfigurator.

I'm not sure where any of this falls in regards to threading. My untrained eye sees that this all seems to be going on within thread 1 - but I have no idea tbh. I'd love to learn about that :)
![Screen Shot 2021-08-06 at 12 05 51 PM](https://user-images.githubusercontent.com/4523694/128546922-33db61ef-41fd-4813-9210-19c6eb83fa0b.png)

Reply:
I removed the printfs, and added a couple logs, both info and debug